### ISSUE TITLE
Python 3 bytes/string error 

### ISSUE THREAD

**Issue by Erotemic** [pygraphviz/pygraphviz#117]
There seems to be an error happening when I'm running agraph, but it seems like the handling of whatever error that is is bugged. 

```
  File "/home/joncrall/code/plottool/plottool/nx_helpers.py", line 714, in nx_agraph_layout
    agraph.layout(prog='neato', args='-n ' + args)
  File "/home/joncrall/venv3/lib/python3.5/site-packages/pygraphviz/agraph.py", line 1358, in layout
    data = self._run_prog(prog, ' '.join([args, "-T", fmt]))
  File "/home/joncrall/venv3/lib/python3.5/site-packages/pygraphviz/agraph.py", line 1338, in _run_prog
    warnings.warn(b"".join(errors), RuntimeWarning)
TypeError: cannot use a string pattern on a bytes-like object
```

I dropped into where the error happened and the errors list looks like this: 
```
errors = [b'Warning: (0.98135718000000005, 0.19697982, 0.021173419999999998, 1.0) is not a known color.\nWarning: (1.0, 0.45278534999999998, 0.0, 1.0) is not a known color.\nWarning: (1.0, 0.33971852000000002, 0.0, 1.0) is not a known color.\nWarning: (1.0, 0.65106306000000003, 0.0, 1.0) is not a known color.\n']

```

Running the line: `b"".join(errors)` seems to work ok, but running `warnings.warn(b"".join(errors), RuntimeWarning)` results in the `TypeError`. I made a quick patch by wrapping the resulting byte object in str which seems to work `warnings.warn(str(b"".join(errors)), RuntimeWarning)`

**Comment by sebastialonso** [comment]
Same thing happening to me. I tested this locally and the layout was calculated fine. But now that I've deployed the service and is receiving real life parameters this bug jumps. I'll try some more things and report back.

**Comment by sebastialonso** [comment]
Ok, I tried to run the service locally and something intereresting happened

Here's the log
~~~
layout_1       | No layouts found. Retrieving network and calculating..
layout_1       | Building layout for network 24 with algorith circo
layout_1       | Building layout for network 24 with algorith neato
layout_1       | Building layout for network 24 with algorith twopi
layout_1       | Building layout for network 24 with algorith fdp
layout_1       | ERROR:tornado.application:Uncaught exception GET /api/v1/layouts/24?algo=dot (172.18.0.1)
layout_1       | HTTPServerRequest(protocol='http', host='localhost:4000', method='GET', uri='/api/v1/layouts/24?algo=dot', version='HTTP/1.1', remote_ip='172.18.0.1', headers={'Host': 'localhost:4000', 'User-Agent': 'curl/7.47.0', 'Authorization': 'Token token=7de279029005f285050777596378907dc20c5326', 'Accept': '*/*'})
layout_1       | Traceback (most recent call last):
layout_1       |   File "/usr/local/lib/python3.5/site-packages/tornado/web.py", line 1425, in _stack_context_handle_exception
layout_1       |     raise_exc_info((type, value, traceback))
layout_1       |   File "<string>", line 3, in raise_exc_info
layout_1       |   File "/usr/local/lib/python3.5/site-packages/tornado/web.py", line 1627, in wrapper
layout_1       |     result = method(self, *args, **kwargs)
layout_1       |   File "/usr/src/app/service/controllers/handlers.py", line 49, in get
layout_1       |     builder.build_layouts_for_network(int(id), self.token)
layout_1       |   File "/usr/src/app/service/builder.py", line 44, in build_layouts_for_network
layout_1       |     layouts = graph.calculate_all_layouts()
layout_1       |   File "/usr/src/app/service/graph.py", line 23, in calculate_all_layouts
layout_1       |     generated_positions = self.algos[algo](self.G)
layout_1       |   File "/usr/src/app/service/algorithms.py", line 7, in <lambda>
layout_1       |     'fdp': lambda G: graphviz_layout(G, prog="fdp"),
layout_1       |   File "/usr/local/lib/python3.5/site-packages/networkx/drawing/nx_agraph.py", line 228, in graphviz_layout
layout_1       |     return pygraphviz_layout(G,prog=prog,root=root,args=args)
layout_1       |   File "/usr/local/lib/python3.5/site-packages/networkx/drawing/nx_agraph.py", line 262, in pygraphviz_layout
layout_1       |     A.layout(prog=prog,args=args)
layout_1       |   File "/usr/local/lib/python3.5/site-packages/pygraphviz/agraph.py", line 1358, in layout
layout_1       |     data = self._run_prog(prog, ' '.join([args, "-T", fmt]))
layout_1       |   File "/usr/local/lib/python3.5/site-packages/pygraphviz/agraph.py", line 1338, in _run_prog
layout_1       |     warnings.warn(b"".join(errors), RuntimeWarning)
layout_1       | TypeError: cannot use a string pattern on a bytes-like object
~~~

If you notice, I build layouts sequentially, and there seems to be problem only with `fdp`.
In effect, if I run the service again (having commented out the fdp option) it works like a charm.

Any ideas?

This was working well yesterday... :(

**Comment by hagberg** [comment]
I think we got that encoding issue  fixed in https://github.com/pygraphviz/pygraphviz/commit/272a113976b809d46d3a1c3a95d00bd82c6c93a8
If that doesn't solve your issue let us know.

**Comment by kdheepak** [comment]
I have the same issue and this fix works for me. This change is not in a stable release yet though. When is a 1.4 stable release expected?

**Comment by marcevrard** [comment]
Using `pygraphviz 1.4rc1` with `python 3.6.2` still yields a similar error:
```
~/anaconda/lib/python3.6/site-packages/pygraphviz/agraph.py in _run_prog(self, prog, args)
   1333         if len(errors) > 0:
   1334             warnings.warn(b"".join(errors).decode(self.encoding), RuntimeWarning)
-> 1335         return b"".join(data)
   1336 
   1337     def layout(self, prog='neato', args=''):
OSError: b''
```

**Comment by lukewitmer** [comment]
This issue is still open. The fix Erotemic mentioned works for me too.

**Comment by marcidy** [comment]
The below is not a proper fix, but this change makes it not fail and returns the RuntimeWarnings on 1.3.1:

line 1338:

    warnings.warn("".join(str(errors)), RuntimeWarning)

I'm not sure why this is different, as this works fine in a shell

    >>> b''.join([b'asdf', b'asdf', b'asdf'])

I need to this to work "right now" so im proceeding with this, but hopefully can come back around shortly to send a pr.

**Comment by andreacassioli** [comment]
Got the same issue. Is fix 272a113 on its way?

**Comment by Geocali** [comment]
I also have a related error : 

TypeError                                 Traceback (most recent call last)
<ipython-input-49-6fefdc498455> in <module>()
      1 B=pgv.AGraph(nvo) # create a new graph from file
----> 2 B.layout(prog='dot') # layout with default (neato)
      3 B.draw('simple.png') # draw png

~/anaconda3/lib/python3.6/site-packages/pygraphviz/agraph.py in layout(self, prog, args)
   1356         """
   1357         fmt = 'dot'
-> 1358         data = self._run_prog(prog, ' '.join([args, "-T", fmt]))
   1359         self.from_string(data)
   1360         self.has_layout = True

~/anaconda3/lib/python3.6/site-packages/pygraphviz/agraph.py in _run_prog(self, prog, args)
   1336 
   1337         if len(errors) > 0:
-> 1338             warnings.warn(b"".join(errors), RuntimeWarning)
   1339 
   1340         return b"".join(data)

TypeError: cannot use a string pattern on a bytes-like object

**Comment by jlumpe** [comment]
@marcidy The error is due to the `b''.join(errors)`, but passing a bytes object to `warn()`.

**Comment by lcwheeler** [comment]
Just had the same error. Added to fixes from [272a113](https://github.com/pygraphviz/pygraphviz/commit/272a113976b809d46d3a1c3a95d00bd82c6c93a8) directly to agraph.py in my installation (which is up to date, the newest version from pip). Those changes fixed the error.

**Comment by jason79** [comment]
I had the same problem. turns out it didn't like `>` in a string which I try to put into an html table. Once removed it was ok.

**Comment by dschult** [comment]
The fix [272a113 ](https://github.com/pygraphviz/pygraphviz/commit/272a113976b809d46d3a1c3a95d00bd82c6c93a8) is in v1.5 and after...
